### PR TITLE
Add support for credential_process

### LINKED ISSUE #1485: Missing support for credential_process
minio-go can obtain credentials from a configured aws config file.

This process, however, only looks for static credentials and does not support the [credential_process](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sourcing-external.html) mechanism to execute an external program that obtains the required credentials.

This is useful in environments where federated authentication is in place and long-living credentials are forbidden.

### PR DESCRIPTION
**Pull request by albertomurillo**
Fixes #1485

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

### CODE REVIEW

File: pkg/credentials/file_aws_credentials.go [hunk 1]
```diff
@@ -18,13 +18,26 @@
...
 	ini "gopkg.in/ini.v1"
 )
 
+// A externalProcessCredentials stores the output of a credential_process
+type externalProcessCredentials struct {
+	Version         int
+	SessionToken    string
+	AccessKeyId     string
+	SecretAccessKey string
+	Expiration      string
```
Comment (harshavardhana) [inline pkg/credentials/file_aws_credentials.go:38]:
```suggestion
	Expiration      time.Time
```

File: pkg/credentials/file_aws_credentials.go [hunk 2]
```diff
@@ -101,7 +146,12 @@ func (p *FileAWSCredentials) Retrieve() (Value, error) {
 
 // IsExpired returns if the shared credentials have expired.
 func (p *FileAWSCredentials) IsExpired() bool {
-	return !p.retrieved
+	if p.expiration.IsZero() {
+		return !p.retrieved
+	}
+
+	now := time.Now()
```
Comment (harshavardhana) [inline pkg/credentials/file_aws_credentials.go:153]: you need to add lee-way you can't expire at the last nano-second @albertomurillo look at iam_aws implementation for ideas.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: pkg/credentials/file_aws_credentials.go (hunk 2 above)
Comment (albertomurillo) [reply pkg/credentials/file_aws_credentials.go:148]: Nice, I updated `FileAWSCredentials` struct to make use of credentials.go `Expiry`.

Now that `DefaultExpiryWindow` is being used in two files, what do you think about refactoring it closer to `Expiry` definition in credentials.go ?

File: pkg/credentials/file_aws_credentials.go [hunk 3]
```diff
@@ -90,6 +105,31 @@ func (p *FileAWSCredentials) Retrieve() (Value, error) {
 	// Default to empty string if not found.
 	token := iniProfile.Key("aws_session_token")
 
+	// If credential_process is defined, obtain credentials by executing
+	// the external process
+	credential_process := iniProfile.Key("credential_process").String()
+	if credential_process != "" {
+		args := strings.Fields(credential_process)
+		cmd := exec.Command(args[0], args[1:]...)
```
Comment (klauspost) [inline pkg/credentials/file_aws_credentials.go:113]: Validate `len(args) > 1` to avoid crash.

File: pkg/credentials/file_aws_credentials.go [hunk 4]
```diff
@@ -90,6 +105,31 @@ func (p *FileAWSCredentials) Retrieve() (Value, error) {
 	// Default to empty string if not found.
 	token := iniProfile.Key("aws_session_token")
 
+	// If credential_process is defined, obtain credentials by executing
+	// the external process
+	credential_process := iniProfile.Key("credential_process").String()
```
Comment (klauspost) [inline pkg/credentials/file_aws_credentials.go:110]: Use camel case:
```suggestion
	credentialProcess := iniProfile.Key("credential_process").String()
```

**Comment by harshavardhana** [comment]
@albertomurillo Please re-submit the PR when it's ready.  Closing this on our end.

**Comment by albertomurillo** [comment]
> @albertomurillo Please re-submit the PR when it's ready. Closing this on our end.

Can’t rework due to AGPL license change.

I hope someone else takes over the patch.

**Pull request closed without merging**